File: gripl/gripl-backend/src/main/kotlin/de/mertendieckmann/griplbackend/model/dto/AnalysisDto.kt
package de.mertendieckmann.griplbackend.model.dto

import de.mertendieckmann.griplbackend.application.BpmnExtractor
import de.mertendieckmann.griplbackend.model.BpmnElement
import de.mertendieckmann.griplbackend.model.analysis.BpmnAnalysisResult

data class AnalysisResponse(
    val criticalElements: List<CriticalElement>,
    val amountOfRetries: Int? = null,
    val ragContext: Map<String, RagElementContext>? = null,
    val ragPromptContext: List<String>? = null
) {
    data class CriticalElement(
        val id: String,
        val name: String?,
        val type: String? = null,
        val reason: String,
        val references: List<LlmReference> = emptyList()
    )

    companion object {
        fun fromBpmnAnalysisResult(
            result: BpmnAnalysisResult,
            bpmnElements: Set<BpmnElement>,
            amountOfRetries: Int,
            ragContext: Map<String, RagElementContext>? = null,
            ragPromptContext: List<String>? = null
        ): AnalysisResponse {
            val elements = result.elements.map { element ->
                val bpmnElement = bpmnElements.find { it.id == element.id }
                CriticalElement(
                    id = element.id,
                    // Prefer the element's real BPMN name, then the deterministic name
                    // derived from its labeled flows; the LLM-generated label is the last
                    // resort for unnamed elements without any labeled flows.
                    name = bpmnElement?.name?.takeIf { it.isNotBlank() }
                        ?: bpmnElement?.derivedNameFromFlows()
                        ?: element.label,
                    type = bpmnElement?.type,
                    reason = element.reason,
                    references = element.references.map { ref ->
                        LlmReference(exactText = ref.exactText, sourceDocument = ref.sourceDocument)
                    }
                )
            }

            return AnalysisResponse(
                criticalElements = elements,
                amountOfRetries = amountOfRetries,
                ragContext = ragContext,
                ragPromptContext = ragPromptContext
            )
        }

        fun fromBpmnAnalysisResult(result: BpmnAnalysisResult, bpmnXml: String, amountOfRetries: Int): AnalysisResponse {
            val extractor = BpmnExtractor()
            val bpmnElements = extractor.extractBpmnElements(bpmnXml)
            return fromBpmnAnalysisResult(result, bpmnElements, amountOfRetries)
        }
    }
}

data class RagElementContext(
    val activityName: String?,
    val entities: List<RagEntity>,
    val relationships: List<RagRelationship>,
    val documents: List<RagDocument>
)

data class RagEntity(
    val label: String,
    val type: String,
    val description: String
)

data class RagRelationship(
    val source: String,
    val target: String,
    val label: String
)

data class LlmReference(
    val exactText: String,
    val sourceDocument: String? = null
)

data class RagDocument(
    val content: String,
    val preview: String,
    val sourceDocument: String? = null
)

File: gripl/gripl-frontend/src/models/dto/AnalysisDto.ts
import {GdprCategory, normalizeFrontendGdprCategories} from "@/models/GdprCategory";

export interface AnalysisRequest {
    bpmnXml: String
}

export interface BinaryAnalysisResponse {
    criticalElements: CriticalElement[];
    ragContext?: Record<string, RagElementContext>;
    amountOfRetries?: number | null;
}

export interface CriticalElement {
    id: string;
    name?: string | null;
    type?: string;
    reason: string;
    references?: LlmReference[];
}

export interface MulticlassAnalysisResponse {
    classifiedElements: ClassifiedElement[];
    amountOfRetries?: number | null;
}

export interface ClassifiedElement {
    id: string;
    name?: string | null;
    reason: string;
    classification: GdprCategory[];
}

export type AnalysisResponse = BinaryAnalysisResponse | MulticlassAnalysisResponse;

export interface NormalizedAnalysisElement {
    id: string;
    name: string;
    type?: string;
    reason: string;
    references?: LlmReference[];
    classification: GdprCategory[];
}

export interface LlmReference {
    exactText: string;
    sourceDocument?: string | null;
}

export interface RagElementContext {
    activityName: string | null;
    entities: RagEntity[];
    relationships: RagRelationship[];
    documents: RagDocument[];
}

export interface RagEntity {
    label: string;
    type: string;
    description: string;
}

export interface RagRelationship {
    source: string;
    target: string;
    label: string;
}

export interface RagDocument {
    content: string;
    preview: string;
    sourceDocument?: string | null;
}

export interface PdfHighlightRect {
    x0: number;
    y0: number;
    x1: number;
    y1: number;
}

export interface PdfLocateResponse {
    page: number | null;
    rects: PdfHighlightRect[];
    page_width: number;
    page_height: number;
}

export function getAnalysisElements(response: AnalysisResponse | null | undefined): NormalizedAnalysisElement[] {
    if (!response) {
        return [];
    }

    if ("classifiedElements" in response) {
        return response.classifiedElements.map((element) => ({
            id: element.id,
            name: element.name || element.id,
            reason: element.reason,
            classification: normalizeFrontendGdprCategories(element.classification as unknown as string[]),
        }));
    }

    return response.criticalElements.map((element) => ({
        id: element.id,
        name: element.name || element.id,
        reason: element.reason,
        classification: [],
    }));
}